packages/preact/src/search/searchWithCachePrefill.ts

```diff
@@ -0,0 +1,96 @@
+import { search } from "@core/search"
+import { SearchOptions } from "@core/types"
+import { SearchQuery, SearchResult } from "@nosto/nosto-js/client"
+import { Config } from "@preact/config/config"
+import { deepMerge } from "@utils/deepMerge"
+
+import { applyQueryDefaults } from "./defaults"
+import { cacheSearchResult, loadCachedResultIfApplicable } from "./resultCaching"
+
+type CacheContextOptions = {
+  cacheResult?: SearchResult
+  prefillQuery?: SearchQuery
+}
+
+export async function searchWithCache(config: Config, searchQuery: SearchQuery, options: SearchOptions) {
+  const usePersistentCache = config.pageType !== "autocomplete" && config.persistentSearchCache
+
+  const { cacheResult, prefillQuery } = getCacheResult(usePersistentCache, searchQuery) || {}
+
+  if (cacheResult) {
+    if (!prefillQuery) {
+      return cacheResult
+    }
+    const backfillResponse = await performSearch(config, options, prefillQuery)
+    const mergedResponse = {
+      ...cacheResult,
+      products: {
+        ...cacheResult.products,
+        hits: [...(cacheResult.products?.hits || []), ...(backfillResponse.products?.hits || [])],
+        total: backfillResponse.products?.total
+      }
+    }
+    cacheSearchResult(usePersistentCache, searchQuery, mergedResponse)
+    return mergedResponse
+  }
+
+  const response = await performSearch(config, options, searchQuery)
+  cacheSearchResult(usePersistentCache, searchQuery, response)
+  return response
+}
+
+function getCacheResult(usePersistentCache: boolean, searchQuery: SearchQuery): CacheContextOptions | undefined {
+  const { from, size = 0 } = searchQuery.products || {}
+
+  const cacheData = loadCachedResultIfApplicable(usePersistentCache, searchQuery)
+
+  // no cache data available
+  if (!cacheData) {
+    return undefined
+  }
+
+  const { from: cacheFrom = 0, size: cacheSize = 0 } = cacheData.query.products || {}
+
+  // a regular pagination query
+  if (from && from > 0) {
+    if (from === cacheFrom && size === cacheSize) {
+      return { cacheResult: cacheData.result }
+    }
+    return undefined
+  }
+
+  // cache size matches the requested size
+  // works for the first loaded result with all render modes (from = 0, size = DEFAULT_SIZE)
+  if (size === cacheSize) {
+    return { cacheResult: cacheData.result }
+  }
+
+  // requested size is more than the cache size. So, use cache data for prefilling
+  const prefillFrom = size - cacheSize
+  const prefillSize = size - prefillFrom
+
+  return {
+    cacheResult: cacheData.result,
+    prefillQuery:
```

Comment:
why not return the result directly?

Reply:
Not sure if I understand the comment correctly. Can you elaborate?